File: Pie02/Pie02CS/Expressions/NamespaceDeclaration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Irony.Parsing;

namespace Pie02CS.Expressions
{
    class NamespaceDeclaration
        :Expression
    {
        public NamespaceDeclaration(Expression parentExpression)
            : base(parentExpression)
        {

        }

        public static void Build(IronyParser parser, Expression parentExpresion, ParseTreeNode node)
        {
            var lastEx = parentExpresion;
            foreach(var name in node.ChildNodes[1].ChildNodes)
            {
                var n = new NamespaceDeclaration(lastEx);
                n.Name = name.FindTokenAndGetText();
                lastEx.Children.Add(n);
                lastEx = n;
            }
            parser.ConsumeParseTree(lastEx, node.ChildNodes[2]);
        }

        public override void Emit(ref string source)
        {
            source += " namespace ";
            source += Name;
            source += "\n{\n";

            base.Emit(ref source);

            source += "}\n";
        
        }

        public override List<dynamic> Match(Expression ex)
        {
            if (ex is NamespaceDeclaration)
                if (ex.GetFullName() == GetFullName())
                    return new List<dynamic>() { this };
            return base.Match(ex);
        }

        public override string GetFullName()
        {
            if (ParentExpression == null)
                return Name;
            else
            {
                if (ParentExpression is NamespaceDeclaration)
                {
                    return ParentExpression.GetFullName() + "." + Name;
                }
                else return Name;
            }
        }

        public override void Validate()
        {


            base.Validate();
        }
    }
}

File: Pie02/Pie02IronPython/Expressions/NamespaceDeclaration.py
from Expression import Expression

class NamespaceDeclaration(Expression):
    """description of class"""
    def __init__(self, parentExpression):
        super(self.__class__, self).__init__(parentExpression)



    @staticmethod
    def Build(parser, parent, node):
        lastEx = parent
        for name in node.ChildNodes[1].ChildNodes:
            n = NamespaceDeclaration(parent)
            n.Name = name.FindTokenAndGetText()
            lastEx.Children.append(n)
            lastEx = n
        parser.ConsumeParseTree(lastEx, node.ChildNodes[3])    
        
          

    def Emit(self, source):
        source += "namespace "
        source += self.Name
        source += "\n{\n"

        source = super(self.__class__, self).Emit(source)

        source += "}\n"
        return source

    def Match(self, ex):
        print isinstance(ex, NamespaceDeclaration)
        if type(ex) is NamespaceDeclaration:
            print "validating namespace " + self.Name + " " + ex.Name
            if ex.GetFullName() == self.GetFullName():
                return self
        return super(self.__class__, self).Match(ex)

    def GetFullName(self):
        if self.ParentExpression is None:
            return self.Name
        else:
            if type(self.ParentExpression) is Expression:
                return self.Name
            else:
                return self.ParentExpression.Name + "." + self.Name